filecoin-proofs/src/api/post_adapter.rs

```diff
@@ -0,0 +1,626 @@
+            comm_rs[i] = *comm_r;
+        }
+
+        fixed.push(VerifyPoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            comm_rs,
+            challenge_seed: dynamic.challenge_seed,
+            proof: dynamic.proofs[i],
+            faults: Vec::new(),
+        })
+    }
+
+    // If we receive a number of comm_rs which does not divide evenly by
+    // POST_SECTORS_COUNT, form a grouping of comm_rs which does divide
+    // evenly by POST_SECTORS_COUNT by duplicating the last comm_r in the
+    // group until LEN(comm_rs) == POST_SECTORS_COUNT.
+    if !remainder.is_empty() {
+        let mut comm_rs: [Commitment; POST_SECTORS_COUNT] = Default::default();
+
+        let vs = repeat(remainder[remainder.len() - 1]);
+        let rs = remainder.iter().cloned();
+        let it = rs.chain(vs).take(POST_SECTORS_COUNT).enumerate();
+
+        for (i, comm_r) in it {
+            comm_rs[i] = comm_r;
+        }
+
+        fixed.push(VerifyPoStFixedSectorsCountInput {
+            sector_bytes: dynamic.sector_bytes,
+            comm_rs,
+            challenge_seed: dynamic.challenge_seed,
+            proof: dynamic.proofs[proofs_len - 1],
+            faults: Vec::new(),
+        })
+    }
+
+    // Map the dynamic sectors count indices to the fixed sectors count indices.
+    for fault in dynamic.faults {
+        let i = (fault as f64 / POST_SECTORS_COUNT as f64).floor();
+        fixed[i as usize]
+            .faults
+            .push(fault % POST_SECTORS_COUNT as u64);
+    }
+
+    // If we had to duplicate commitments to make LEN(COMMITMENTS) ==
+    // POST_SECTORS_COUNT and the thing we duplicated was faulty, ensure that
+    // the duplicates are marked as faulty, too.
+    if !remainder.is_empty() {
+        let fixed_len = { fixed.len() - 1 };
+        let remdr_len = { remainder.len() };
+
+        let last_remainder_fault: Option<u64> =
+            { fixed[fixed_len].faults.get(remdr_len - 1).cloned() };
+
+        if let Some(fault) = last_remainder_fault {
+            for n in 0..(POST_SECTORS_COUNT - remdr_len) {
+                fixed[fixed_len].faults.push(1 + fault + n as u64);
+            }
+        }
+    }
+
+    Ok(fixed)
+}
+
+/// Collapses return values from multiple fixed sectors-count PoSt verify
+/// operations into a single return value for dynamic sector count.
+///
+pub fn verify_post_collect_output(
+    xs: Vec<error::Result<VerifyPoStFixedSectorsCountOutput>>,
+) -> error::Result<VerifyPoStDynamicSectorsCountOutput> {
+    if xs.is_empty() {
+        return Err(format_err!("input vector must not be empty"));
+    }
+
+    let mut dynamic = VerifyPoStDynamicSectorsCountOutput { is_valid: true };
+
+    for x in xs.into_iter() {
+        match x {
+            Err(e) => {
+                return Err(e);
+            }
+            Ok(VerifyPoStFixedSectorsCountOutput { is_valid }) => {
+                dynamic.is_valid = dynamic.is_valid && is_valid;
```

Comment:
Seems we can just stop here and return a `false` the first time it happens.

Reply:
I rewrote the reduction in a functional style. The enclosing function now has a single exit point instead of multiple exit points, which I think makes it clearer. Thanks for calling this out.